src/java/io/bazel/rulesscala/scalac/reporter/DepsTrackingReporter.java

```diff
@@ -0,0 +1,330 @@
+            reporter.warning(NoPosition$.MODULE$, compilerDepsReport.toString());
+          } else {
+            reporter.error(NoPosition$.MODULE$, compilerDepsReport.toString());
+          }
+        }
+      }
+
+      if (!ops.unusedDependencyCheckerMode.equals("off")) {
+        StringBuilder unusedDepsReport = new StringBuilder("Unused dependencies:\n");
+        int count = 0;
+        for (Dep dep : unusedDeps) {
+          if (dep.ignored) {
+            continue;
+          }
+          count++;
+          unusedDepsReport.append(reportDep(dep, false));
+        }
+        if (count > 0) {
+          if (ops.unusedDependencyCheckerMode.equals("warn")) {
+            reporter.warning(NoPosition$.MODULE$, unusedDepsReport.toString());
+          } else if (ops.unusedDependencyCheckerMode.equals("error")) {
+            reporter.error(NoPosition$.MODULE$, unusedDepsReport.toString());
+          }
+        }
+      }
+    }
+  }
+
+  private String reportDep(Dep dep, boolean add) {
+    if (add) {
+      String message =
+          "error: Target '" + dep.target + "' (via jar: ' " + dep.jar + " ') is being used by "
+              + ops.currentTarget
+              + " but is is not specified as a dependency, please add it to the deps.\nYou can use the following buildozer command:\n";
+      String command = "buildozer 'add deps " + dep.target + "' " + ops.currentTarget;
+      return message + command + "\n";
+    } else {
+      String message = "error: Target '" + dep.target + "' (via jar: ' " + dep.jar
+          + " ')  is specified as a dependency to " + ops.currentTarget
+          + " but isn't used, please remove it from the deps.\nYou can use the following buildozer command:\n";
+      String command = "buildozer 'remove deps " + dep.target + "' " + ops.currentTarget;
+      return message + command + "\n";
+    }
+  }
+
+  private String guessFullJarPath(String jar) {
+    if (jar.endsWith(IJAR_JAR_SUFFIX)) {
+      return stripIjarSuffix(jar, IJAR_JAR_SUFFIX);
+    } else if (jar.endsWith(HJAR_JAR_SUFFIX)) {
+      return stripIjarSuffix(jar, HJAR_JAR_SUFFIX);
+    } else {
+      return jar;
+    }
+  }
+
+  private static String stripIjarSuffix(String jar, String suffix) {
+    return jar.substring(0, jar.length() - suffix.length()) + ".jar";
+  }
+
+  private String jarLabel(String path) {
+    try (JarFile jar = new JarFile(path)) {
+      return jar.getManifest().getMainAttributes().getValue("Target-Label");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void registerAstUsedJars(Set<String> jars) {
+    astUsedJars = jars;
+  }
+
+  public void writeDiagnostics(String diagnosticsFile) throws IOException {
+    if (delegateReporter == null) {
+      return;
+    }
+
+    ProtoReporter protoReporter = (ProtoReporter) delegateReporter;
+    protoReporter.writeTo(Paths.get(diagnosticsFile));
+  }
+
+  private static class Dep {
```

Comment:
Is this class really needed. It looks like it maps one to one to `Dependencies` from proto. You could cut some conversion code. But I myself not convinced this would be proper approach.

Reply:
Replaced with `Dependencies`